### DISCUSSION TITLE
Difference between x and X

### DISCUSSION THREAD

**Discussion by yusuphgammut** [helix-editor/helix#6031] [Q&A] [answered]
Hi folks.

I've been working a bit with Helix and I still can't tell the difference between `x` (`extend_line_below`) and `X` (`extend_to_line_bounds`) other than just extending to the next line in the case of `x`. Is there a use case were any of you use `X` which can't be resolved using `x` just once? I mean, if I can press `x` one time and it's the same as pressing `X`, there's no reason to have `X` IMHO.

**Comment by CptPotato** [comment, marked as answer]
One difference is that `extend_line_below` will extend to the next line if your cursor is on an empty line (since the whole line is already "selected").

**Reply by CptPotato** [reply]
i.e. if you want to delete the current line, `Xd` will always work. With `xd` on the other hand you may accidentally delete two lines if your cursor is on an empty line.

I personally have `x` and `X` swapped in my config for convenience.

**Reply by David-Else** [reply]
But if you want to delete an empty line just press d? I don't really get the use case for `x` vs `X`, do you have any more examples?

Regarding your swapping, do you not find yourself using `X` a lot more, I would imagine I would, but maybe I should use the `[` `]` selectors more...

**Reply by CptPotato** [reply]
My reasoning is that I can always delete a line with `xd` (using **my swapped bindings**) regardless of the lines content.

> But if you want to delete an empty line just press d?

If a line happens to contain whitespace (such as indentation) this will not work. I also don't like the inconsistency of doing this differently depending on the line's content.

> Regarding your swapping, do you not find yourself using `X` a lot more [...]

I use both frequently. But usually when I extend I do so multiple times, so I can just hold the <kbd>Shift</kbd> with my bindings. And `xd` for a single line is much less awkward than the default bindings `Xd`.

(also see #835)

**Reply by David-Else** [reply]
Good points, I will have to experiment. I try to stick to the defaults where I can as I assume the person who made them has some knowledge I don't and there is a way of doing things that I don't yet know about... apart from these non negotiable essentials :)

```
[keys.normal]
G = "goto_file_end"          # vim
Z = { Z = ":wq", Q = ":q!" } # vim, save and quit, quit without saving
"#" = "toggle_comments"

[keys.insert]
j = { k = "normal_mode" }
```

**Reply by yusuphgammut** [reply]
Thanks for the answers. I now have at least 1 fairly valid use case for `X`.

I did notice that there's a chance that the `x` and `X` keymaps change in the future if a more robust visual/select mode is implemented as per this [wiki page](https://github.com/helix-editor/helix/wiki/Keymap-brainstorm). I miss a lot the line-wise visual selection in Vim where you can extend up or down using `k` and `j`. Helix's `x` only flowing downwards seems a bit restrictive to me but I think I should get used to use more complex object selection.
